### PR TITLE
SPY-885: CSD's FileSelector is not applied for Parquet in SparkSQL query

### PR DESCRIPTION
**Pull request by ianlcsd**
@markhamstra 
@mbautin

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 1]
```diff
@@ -508,7 +507,9 @@ private[hive] class HiveMetastoreCatalog(val client: ClientInterface, hive: Hive
 
       parquetRelation
     } else {
-      val paths = Seq(metastoreRelation.hiveQlTable.getDataLocation.toString)
+      val paths = selectParquetLocationDirectories(
+          metastoreRelation.tableName,
+          Option(metastoreRelation.hiveQlTable.getDataLocation.toString))
```
Comment (mbautin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Code style: indent lines 511-512 only two spaces over line 510.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 2]
```diff
@@ -524,6 +525,36 @@ private[hive] class HiveMetastoreCatalog(val client: ClientInterface, hive: Hive
...
+   * Customizing the data directory selection by using hadoopFileSelector.
+   *
+   * The value of locationOpt will be included and returned if
+   * 1. the hadoopFileSelector is not defined or
+   * 2. locationOpt is not defined or
+   * 3. the selected directories is empty.
+   *
+   * Otherwise, the none empty selected directories will be returned.
+   */
+  def selectParquetLocationDirectories(
```
Comment (mbautin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Can this be protected / private?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 3]
```diff
@@ -524,6 +525,36 @@ private[hive] class HiveMetastoreCatalog(val client: ClientInterface, hive: Hive
...
 
+  /**
+   * Customizing the data directory selection by using hadoopFileSelector.
+   *
+   * The value of locationOpt will be included and returned if
+   * 1. the hadoopFileSelector is not defined or
+   * 2. locationOpt is not defined or
+   * 3. the selected directories is empty.
+   *
+   * Otherwise, the none empty selected directories will be returned.
```
Comment (mbautin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Did you mean "non-empty"?

**Comment by ianlcsd** [comment]
having Jenkins problem now: "Detected Maven Version: 3.0.5 is not in the allowed range 3.3.3."

**Comment by markhamstra** [comment]
What are you trying to do with Jenkins?  Spark needs to build with maven 3.3.3+, which isn't the default on Jenkins (because the deploy plugin doesn't work correctly with recent maven versions -- don't get me started....)  Spark builds should use the `build/mvn` script from the Spark tree instead of using maven directly.  Add `--force` as the first option to `build/mvn`, and an appropriate maven version will be downloaded and used -- all other usual mvn options follow that.

**Comment by ianlcsd** [comment]
Jenkins, test this please.

**Comment by ianlcsd** [comment]
Jenkins, test this please.

**Comment by ianlcsd** [comment]
Jenkins, test this please.

**Comment by ianlcsd** [comment]
Jenkins, test this please.

**Comment by ianlcsd** [comment]
Jenkins, test this please.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 4]
```diff
@@ -45,6 +41,9 @@ import org.apache.spark.sql.sources._
 import org.apache.spark.sql.types._
 import org.apache.spark.sql.{AnalysisException, SQLContext, SaveMode}
 
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+
```
Comment (markhamstra) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Don't re-order the imports.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 5]
```diff
@@ -524,6 +525,36 @@ private[hive] class HiveMetastoreCatalog(val client: ClientInterface, hive: Hive
...
+      selector <- hive.hadoopFileSelector
+      l <- locationOpt
+      location = new Path(l)
+      fs = location.getFileSystem(hive.hiveconf)
+      selectedPaths <- selector.selectFiles(tableName, fs, location)
+      selectedDir = for {
+        selectedPath <- selectedPaths
+        if selectedPath.getFileSystem(hive.hiveconf).isDirectory(selectedPath)
+      } yield selectedPath.toString
+      if !selectedDir.isEmpty
```
Comment (markhamstra) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]:
```scala
} yield selectedPath.toString if selectedDir.nonEmpty
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 6]
```diff
@@ -524,6 +525,36 @@ private[hive] class HiveMetastoreCatalog(val client: ClientInterface, hive: Hive
...
+    } yield selectedDir
+    inputPaths.getOrElse(Seq(locationOpt.getOrElse(null)))
```
Comment (markhamstra) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]:
```scala
inputPaths.getOrElse(Seq(locationOpt.orNull))
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 7]
```diff
@@ -158,5 +241,6 @@ class DataSourceWithHiveMetastoreCatalogSuite extends DataSourceTest with SQLTes
         }
       }
     }
+
```
Comment (markhamstra) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: revert

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 8]
```diff
@@ -50,6 +51,88 @@ class HiveMetastoreCatalogSuite extends SparkFunSuite with Logging {
...
+    assertResult(Seq("somewhere"))(
+      hmc.selectParquetLocationDirectories("sometable", Option("somewhere"))
+    )
+    // none selected
+    assertResult(Seq(null))(
+      hmc.selectParquetLocationDirectories("sometable", Option(null))
+    )
+  }
+
+
```
Comment (markhamstra) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: delete the extra two empty lines

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 9]
```diff
@@ -50,6 +51,88 @@ class HiveMetastoreCatalogSuite extends SparkFunSuite with Logging {
...
+  val baseDir = {
+    val base = File.createTempFile("selectParquetLocationDirectories", "1", hiveFilesTemp)
+    base.delete()
+    base.mkdirs()
+    base
+  }
+
+  test(s"With Selector selecting from ${baseDir.toString}") {
+    val fullpath = (somewhere: String, sometable: String)
+    => s"${baseDir.toString}/${somewhere}/$sometable"
```
Comment (markhamstra) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: Spark Style Guide:
```scala
  val fullpath = { (somewhere: String, sometable: String) =>
    s"${baseDir.toString}/${somewhere}/$sometable"
  }
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 10]
```diff
@@ -50,6 +51,88 @@ class HiveMetastoreCatalogSuite extends SparkFunSuite with Logging {
...
+
+    TestHive.setHadoopFileSelector(new HadoopFileSelector() {
+      override def getFilesSizeInBytes(tableName: String, fs: FileSystem, basePath: Path):
+      Option[Long] = Some(100L)
+
+      override def selectFiles(sometable: String, fs: FileSystem, somewhere: Path):
+      Option[Seq[Path]] =
+        Some(Seq(new Path(fullpath(somewhere.toString, sometable))))
```
Comment (markhamstra) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]:
```scala
      override def getFilesSizeInBytes(
          tableName: String,
          fs: FileSystem,
          basePath: Path): Option[Long] = Some(100L)

      override def selectFiles(
          sometable: String,
          fs: FileSystem,
          somewhere: Path): Option[Seq[Path]] = {
        Some(Seq(new Path(fullpath(somewhere.toString, sometable))))
      }
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 11]
```diff
@@ -50,6 +51,88 @@ class HiveMetastoreCatalogSuite extends SparkFunSuite with Logging {
...
+    })
+
+    // ensure directory existence for somewhere/sometable
+    val somewhereSometable = new File(fullpath("somewhere", "sometable"))
+    somewhereSometable.mkdirs()
+    // somewhere/sometable is a directory => will be selected
+    assertResult(Seq(fullpath("somewhere", "sometable")))(
```
Comment (markhamstra) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]: nit: preference in the codebase for `assertResult` that extends over more than one line appears to be:
```scala
assertResult(foo) {
  .
  .
  .
}
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala [hunk 12]
```diff
@@ -50,6 +51,88 @@ class HiveMetastoreCatalogSuite extends SparkFunSuite with Logging {
...
+    )
+  }
+
+  test("With Selector selecting None") {
+    TestHive.setHadoopFileSelector(new HadoopFileSelector() {
+      override def getFilesSizeInBytes(tableName: String, fs: FileSystem, basePath: Path):
+      Option[Long] = Some(100L)
+      override def selectFiles(tableName: String, fs: FileSystem, basePath: Path):
+      Option[Seq[Path]] =
+        None
```
Comment (markhamstra) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala]:
```scala
Option[Seq[Path]] = None
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 13]
```diff
@@ -524,6 +526,36 @@ private[hive] class HiveMetastoreCatalog(val client: ClientInterface, hive: Hive
     result.copy(expectedOutputAttributes = Some(metastoreRelation.output))
   }
 
+  /**
+   * Customizing the data directory selection by using hadoopFileSelector.
+   *
+   * The value of locationOpt will be returned as single element sequence if
+   * 1. the hadoopFileSelector is not defined or
+   * 2. locationOpt is not defined or
+   * 3. the selected directories is empty.
```
Comment (markhamstra) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: ...selected directories are empty.

**Pull request merged**